File: test/helpers/crawler-fakes.js
/** Mock website (injected fetch) and in-memory crawl store (port of tests/crawler_fakes.py). */

import { createHash } from 'node:crypto';
import { urlKey } from '../../src/utils/urls.js';
import { DatabaseError } from '../../src/utils/errors.js';
import { Fetcher, fetcherConfig } from '../../src/crawler/fetcher.js';
import { SiteCrawler, crawlConfig, pageUpsert } from '../../src/crawler/service.js';

export const BASE = 'https://example.com';
export const SITE_ID = '00000000-0000-0000-0000-0000000c2a71';
export const ROBOTS_ALLOW_ALL = 'User-agent: *\nDisallow:\n';

export function htmlPage(title, body, { canonical = null, headExtra = '', nav = '<a href="/">Home</a>', footer = '© Example' } = {}) {
  const canon = canonical ? `<link rel="canonical" href="${canonical}">` : '';
  return (
    `<!doctype html><html lang="en"><head><title>${title}</title>${canon}${headExtra}</head>` +
    `<body><header><nav>${nav}</nav></header><main>${body}</main>` +
    `<footer>${footer}</footer></body></html>`
  );
}

const NULL_BODY = new Set([101, 103, 204, 205, 304]);

/** A connection-level failure like undici raises (mapped to "ConnectError"). */
export function connectError(message = 'connection reset') {
  return new TypeError('fetch failed', { cause: Object.assign(new Error(message), { code: 'ECONNREFUSED' }) });
}

/** Routes keyed by path (+query). Values: [status, body, headers] or a responder function. */
export class MockSite {
  constructor() {
    this.routes = new Map();
    this.requests = [];
  }

  html(path, title, body, options = {}) {
    this.routes.set(path, [200, htmlPage(title, body, options), { 'content-type': 'text/html' }]);
  }

  add(path, status, body = '', headers = {}) {
    this.routes.set(path, [status, body, Object.fromEntries(Object.entries(headers).map(([k, v]) => [k.replaceAll('_', '-'), v]))]);
  }

  fetch = async (url) => {
    this.requests.push(url);
    const u = new URL(url);
    const route = this.routes.get(u.pathname + u.search) ?? this.routes.get(u.pathname);
    if (route === undefined) {
      return new Response(Buffer.from('not found'), { status: 404, headers: { 'content-type': 'text/html' } });
    }
    if (typeof route === 'function') return route({ url });
    const [status, body, headers] = route;
    const bytes = typeof body === 'string' ? Buffer.from(body, 'utf8') : Buffer.from(body);
    return new Response(NULL_BODY.has(status) ? null : bytes, { status, headers });
  };

  requestedPaths() {
    return this.requests.map((r) => {
      const u = new URL(r);
      return u.pathname + u.search;
    });
  }
}

export class FakeCrawlStore {
  constructor() {
    this.pages = new Map();
    this.upserts = 0;
    this.statusUpdates = [];
    this.failOn = new Set();
    this.locked = false;
  }

  async existingUrls() {
    return new Map([...this.pages.keys()].map((u) => [urlKey(u), u]));
  }

  async upsertPage(siteId, page) {
    if (this.failOn.has(page.url)) {
      throw new DatabaseError(Object.assign(new Error('db down'), { code: '08006' }));
    }
    this.pages.set(page.url, page);
    this.upserts += 1;
    return createHash('sha1').update(page.url).digest('hex');
  }

  async updateStatus(siteId, url, { httpStatus, redirectUrl }) {
    if (!this.pages.has(url)) return false;
    this.pages.set(url, { ...this.pages.get(url), http_status: httpStatus, redirect_url: redirectUrl });
    this.statusUpdates.push([url, httpStatus, redirectUrl]);
    return true;
  }

  async rollback() {}

  async siteLock(siteId, fn) {
    this.locked = true;
    try {
      return await fn();
    } finally {
      this.locked = false;
    }
  }
}

export function makeSite(baseUrl = BASE) {
  return { id: SITE_ID, name: 'Example', base_url: `${baseUrl}/` };
}

export function makeCrawler(mock, store = null, config = {}) {
  const crawlStore = store ?? new FakeCrawlStore();
  const sleeps = [];
  const { maxRetries, maxRedirects, ...rest } = config;
  const fetcher = new Fetcher(
    mock.fetch,
    fetcherConfig({
      userAgent: 'SEOLinkAutomationBot/0.1',
      maxResponseBytes: rest.maxResponseBytes ?? 5_000_000,
      ...(maxRetries !== undefined ? { maxRetries } : {}),
      ...(maxRedirects !== undefined ? { maxRedirects } : {}),
    }),
    { sleep: (s) => sleeps.push(s) },
  );
  const crawler = new SiteCrawler(crawlStore, fetcher, crawlConfig({ requestDelaySeconds: 0.0, ...rest }), {
    sleep: (s) => sleeps.push(s),
  });
  return [crawler, crawlStore, sleeps];
}

/** CrawlRequest defaults as Fastify/Pydantic apply them. */
export function crawlRequest(overrides = {}) {
  return { start_url: null, max_pages: 100, use_sitemaps: true, include_www_variant: false, ...overrides };
}

export const oldPage = (url) => pageUpsert({ url, title: 'old', content_html: '<p>old content</p>' });

File: tests/crawler_fakes.py
"""Mock website (httpx.MockTransport) and in-memory crawl store. No network, no DB."""

from __future__ import annotations

import uuid
from collections.abc import Callable, Iterator
from contextlib import contextmanager
from dataclasses import dataclass, field
from typing import Any

import httpx

from app.core.urls import url_key
from app.crawler.fetcher import Fetcher, FetcherConfig
from app.crawler.service import CrawlConfig, SiteCrawler
from app.pages.models import Site
from app.pages.schemas import PageUpsert

BASE = "https://example.com"
SITE_ID = uuid.UUID("00000000-0000-0000-0000-0000000c2a71")

Responder = Callable[[httpx.Request], httpx.Response]


def html_page(
    title: str,
    body: str,
    *,
    canonical: str | None = None,
    head_extra: str = "",
    nav: str = '<a href="/">Home</a>',
    footer: str = "© Example",
) -> str:
    canon = f'<link rel="canonical" href="{canonical}">' if canonical else ""
    return (
        f'<!doctype html><html lang="en"><head><title>{title}</title>{canon}{head_extra}</head>'
        f"<body><header><nav>{nav}</nav></header><main>{body}</main>"
        f"<footer>{footer}</footer></body></html>"
    )


@dataclass
class MockSite:
    """Routes keyed by path (+query). Values: (status, body, headers) or a responder."""

    routes: dict[str, tuple[int, str | bytes, dict[str, str]] | Responder] = field(
        default_factory=dict
    )
    requests: list[str] = field(default_factory=list)

    def html(self, path: str, title: str, body: str, **kwargs: Any) -> None:
        self.routes[path] = (200, html_page(title, body, **kwargs), {"content-type": "text/html"})

    def add(self, path: str, status: int, body: str | bytes = "", **headers: str) -> None:
        self.routes[path] = (status, body, {k.replace("_", "-"): v for k, v in headers.items()})

    def handler(self, request: httpx.Request) -> httpx.Response:
        self.requests.append(str(request.url))
        key = request.url.raw_path.decode()
        route = self.routes.get(key) or self.routes.get(request.url.path)
        if route is None:
            return httpx.Response(404, text="not found", headers={"content-type": "text/html"})
        if callable(route):
            return route(request)
        status, body, headers = route
        content = body.encode() if isinstance(body, str) else body
        return httpx.Response(status, content=content, headers=headers)

    def client(self) -> httpx.Client:
        return httpx.Client(transport=httpx.MockTransport(self.handler), follow_redirects=False)

    def requested_paths(self) -> list[str]:
        return [httpx.URL(u).raw_path.decode() for u in self.requests]


class FakeCrawlStore:
    def __init__(self) -> None:
        self.pages: dict[str, PageUpsert] = {}
        self.upserts = 0
        self.status_updates: list[tuple[str, int, str | None]] = []
        self.fail_on: set[str] = set()
        self.locked = False

    def existing_urls(self, site_id: uuid.UUID) -> dict[str, str]:
        return {url_key(u): u for u in self.pages}

    def upsert_page(self, site_id: uuid.UUID, page: PageUpsert) -> uuid.UUID:
        if page.url in self.fail_on:
            from sqlalchemy.exc import OperationalError

            raise OperationalError("INSERT", {}, Exception("db down"))
        self.pages[page.url] = page
        self.upserts += 1
        return uuid.uuid5(uuid.NAMESPACE_URL, page.url)

    def update_status(
        self, site_id: uuid.UUID, url: str, *, http_status: int, redirect_url: str | None
    ) -> bool:
        if url not in self.pages:
            return False
        self.pages[url] = self.pages[url].model_copy(
            update={"http_status": http_status, "redirect_url": redirect_url}
        )
        self.status_updates.append((url, http_status, redirect_url))
        return True

    def rollback(self) -> None:
        pass

    @contextmanager
    def site_lock(self, site_id: uuid.UUID) -> Iterator[None]:
        self.locked = True
        try:
            yield
        finally:
            self.locked = False


def make_site(base_url: str = BASE) -> Site:
    return Site(id=SITE_ID, name="Example", base_url=f"{base_url}/")


def make_crawler(
    mock: MockSite, store: FakeCrawlStore | None = None, **config: Any
) -> tuple[SiteCrawler, FakeCrawlStore, list[float]]:
    store = store or FakeCrawlStore()
    sleeps: list[float] = []
    fetcher_kwargs = {k: config.pop(k) for k in ("max_retries", "max_redirects") if k in config}
    fetcher = Fetcher(
        mock.client(),
        FetcherConfig(
            user_agent="SEOLinkAutomationBot/0.1",
            max_response_bytes=config.get("max_response_bytes", 5_000_000),
            **fetcher_kwargs,
        ),
        sleep=sleeps.append,
    )
    crawl_config = CrawlConfig(**{"request_delay_seconds": 0.0, **config})
    return SiteCrawler(store, fetcher, crawl_config, sleep=sleeps.append), store, sleeps


ROBOTS_ALLOW_ALL = "User-agent: *\nDisallow:\n"
